### PR TITLE
feat(tls-route): add TLS route matching crate

### PR DESCRIPTION
**Pull request by zaharidichev**
This PR includes a nonfunctional change that adds a TLS route-matching library.
The library works similarly to the http-route one but instead of relying on HTTP concepts
such as paths, headers, etc is specific to TLS ones (i.e. SNI).

At the moment the library is capable of matching routes based on SNIs.

### CODE REVIEW

File: linkerd/tls/route/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+#![forbid(unsafe_code)]
+
+use linkerd_tls::ServerName;
+use r#match::SessionMatch;
+use tracing::trace;
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct SessionInfo {
+    pub sni: ServerName,
+}
```
Comment (olix0r) [inline linkerd/tls/route/src/lib.rs:14]: This should be moved down below the mods with the other structs. I would probably move it to directly about the `find` impl, so that it's defined nearest where it is used.

It would also be good to include a comment a brief comment describing that this is intended to capture the routing metadata of a proxied/unterminated server-side TLS connection.

File: linkerd/tls/route/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    pub rules: Vec<Rule<P>>,
+}
+
+/// Policies for a given set of route matches.
+#[derive(Clone, Debug, Default, Hash, PartialEq, Eq)]
+pub struct Rule<P> {
+    /// A list of session matchers, *any* of which may apply.
+    ///
+    /// The "best" match is used when comparing rules.
+    pub matches: Vec<r#match::MatchSession>,
```
Comment (olix0r) [inline linkerd/tls/route/src/lib.rs:42]: So it seems that we have both SNI matching at the Route level and also the Rule level. If we're going to include MatchSni at the Route level, then we probably want MatchSession to be empty for now. Otherwise, we probably want to omit the sni matching at the Route level.

I don't have enough information to suggest one or the other. I don't really see any harm in it being defined at the Route level (since this matches the [gateway spec](https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha2/tlsroute_types.go#L87)). Given that the spec does not include any rule based matches currently, I'm not sure of a good reason to include SNI matches in rules.

File: linkerd/tls/route/src/sni.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
+use linkerd_dns as dns;
+use linkerd_tls::ServerName;
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub enum MatchSni {
```
Comment (olix0r) [inline linkerd/tls/route/src/sni.rs:5]: It's probably worth documenting SNI matching semantics with some of the Gateway API's [verbiage](https://github.com/kubernetes-sigs/gateway-api/blob/ffe4b631bf3f0333ef6c6fc902cd4755a369bda0/apis/v1alpha2/tlsroute_types.go#L51-L57).

**Review by olix0r — COMMENTED** [review]
This is looking pretty good to me!

It would probably be good to have some tests cases for sessions that do not match any routes. For example, it would be good to explore the corner cases like a suffix match like `*.test.example.com` and a session for `test.example.com`. What other subtleties should we consider?

**Review by olix0r — APPROVED** [review]

**Comment by olix0r** [comment]
When merging, please ensure that the commit message includes an accurate summary of the change like that in the PR description.
